### ISSUE TITLE
issues disappeared from Hubaord

### ISSUE THREAD

**Issue by mderazon** [rauhryan/huboard#580]
I have some issues associated with a milestone that just disappeared completely from Huboard.

I might have clicked archive, don't think so, but i'm not sure.

I can't seem to find a way to bring them back to Huboard. I can see their metadata inside the issue itself on Github. Here's how it looks

```
<!---
@huboard:{"order":366.9921875,"milestone_order":979,"custom_state":""}
-->
```

**Comment by mderazon** [comment]
I think I also changed the MS name somewhere in the process. Can it be related ?

**Comment by mderazon** [comment]
More findings

reopening the issue via Github puts it back to the "Done" column but closing (via Huboard) it makes to disappear again (not immediately, after a page refresh)

**Comment by rauhryan** [comment]
Closed issues that are NOT in the final column

or 

Closed issues in the final column OLDER than 2 weeks


will both be hidden from the task board

**Comment by mderazon** [comment]
Oh, good to know. It was driving me nuts.
> Closed issues in the final column OLDER than 2 weeks

You mean issues created more than 2 weeks ago ?

I understand the reasons behind both of them, but the 2 weeks magic number doesn't work for everyone.

Is there any way to either configure it / workaround it somehow ?

**Comment by discorick** [comment]
That would be issues that were marked as closed 2 Weeks ago. Theres no way at the moment to configure it at the moment, but it's possible we will implement an option to change that time frame in the future.

**Comment by seanlinsley** [comment]
We've had effectively a month-long sprint b/c of vacations and unexpected interruptions, and this "feature" has hidden 50% of our tickets from this sprint.

If @dahlbyk hadn't mentioned this behavior in #669 I would've thought one of these:
- Huboard is broken
  - it "forgot" about the tickets
  - it sent a bad API request to Github that accidentally blew away some ticket statuses
- A team member accidentally archived some tickets

I put feature in scare quotes because I don't understand why it's desirable in the first place. It's magical behavior that's never described to the user. I would vote to remove it entirely.

**Comment by seanlinsley** [comment]
It appears that Huboard doesn't actually remove the label status, it just hides them from the UI.

<img width="624" alt="screen shot 2016-07-09 at 12 14 52 pm" src="https://cloud.githubusercontent.com/assets/688886/16709180/f6aa2362-45ce-11e6-8300-a58f1ecb6f18.png">

To add some context, roughly 5% of all closed tickets have been automatically hidden by a feature we didn't know about and didn't want.

This is especially a big deal for our team because at the end of each sprint we go through Done and do a final production check, as well as any retrospective needed, before archiving the ticket.

**Comment by mderazon** [comment]
I have to agree. After working with Huboard for quite some time, this uncertainty about whether what we see is all there is or not and constantly comparing the issues on Github with the ones on Huboard is one of the biggest points of friction with Huboard.

@rauhryan can you explain again the reason behind this ?

**Comment by discorick** [comment]
@mderazon @seanlinsley, I agree we should look into this again. The thinking behind this behavior is at some point we need to cap how many closed issues we pull from the API, currently its an arbitrary 2 weeks.

A short term fix is we could raise the cap to some higher number like 4-6 weeks, ultimately in the future maybe making it a configurable option with a hard maximum.

**Comment by seanlinsley** [comment]
Is the purpose to improve performance? I'm confused why all other tickets are loaded, but not closed tickets still on the board after a certain date.

**Comment by dahlbyk** [comment]
> Is the purpose to improve performance? I'm confused why all other tickets are loaded, but not closed tickets still on the board after a certain date.

Consider a board that's new to HuBoard: until every existing closed issue has been archived (each archive step requiring a rate-limited GitHub API call, it would appear in the Done column. Without [batch operations](https://github.com/huboard/huboard/issues/542), this seems a non-starter.

I assume the two week interval was based on Ryan's team's needs at the time, and hasn't changed since. I agree that this is problematic, but if we deviate from the existing two-week window, teams that have gotten used to that default will see unexpected behavior ("I thought these had been archived!"). That said, it seems totally reasonable to make the "archive after N days closed" value configurable. We might even be able to figure out a way to default new boards to a larger window without changing existing boards' behavior.

N.B. [`closed_issues`](https://github.com/huboard/huboard-web/blob/bb840bdc8b91b7087527b4d2a1b38b8ce7f7be5f/lib/bridge/github/issues.rb#L66) is where the default `since` value is specified. Passing through an alternate `since` value through our API should be pretty simple. The harder part will be implementing configuration UI/persistence.

**Comment by seanlinsley** [comment]
It looks like the maximum page size is 100, which is already being set. Couldn't you only show the first 100 closed issues and skip the ones after that?

**Comment by dahlbyk** [comment]
> Couldn't you only show the first 100 closed issues and skip the ones after that?

That's likely what we would use as a hard maximum. The balance we need to strike is new-to-HuBoard users seeing dozens of forgotten closed issues which need to be archived one at a time.

On balance, seeing more issues than you want, with a button to hide, is a better situation to be in than missing issues you need, with no recourse. We'll still probably make this configurable, but it seems reasonable to push out an extension of the Closed window sooner.

Thoughts on a safer window size? Four weeks? Six weeks?

**Comment by seanlinsley** [comment]
Does the Github API wrapper you're using handle pagination? If it doesn't, it looks like the existing code will only load 100 records anyway.

Another option would be to stop showing closed tickets in Done that don't have a status, or have the Backlog status. It causes some confusing webhook notifications for people that close tickets from the backlog from inside Huboard's UI. Specifically it sends a Backlog -> Done notification, and then it also has to be archived so it doesn't show up on the board.

**Comment by seanlinsley** [comment]
I just checked: the issues API lets you filter by an array of labels.

**Comment by seanlinsley** [comment]
A ton of previously archived tickets just showed up in Done for us. Is there a PR / commit for the change?

**Comment by seanlinsley** [comment]
Something seems broken. There were 101 tickets in Done: 1 wasn't closed, 100 were from many years ago. After manually archiving the 100 old tickets and reloading the page, only the 1 open ticket remained, even though we had many closed tickets in Done from the current sprint.

Perhaps https://github.com/huboard/huboard-web/commit/040f0529b06ce21bc4a29686e4635af89123644e had unintended consequences?

**Comment by seanlinsley** [comment]
We currently have 127 closed tickets with the Done label, so it's not like there was an errant API request that removed the label for these tickets.

**Comment by discorick** [comment]
Hey @seanlinsley thats really strange, we haven't deployed anything this week - did those issues just appear yesterday ? 

Is the affected repo public by any chance ? I would be interested to take a look at the issue bodies (where we hide the "archived" meta-data")

> Perhaps huboard/huboard-web@040f052 had unintended consequences?

Shouldn't, that commit just filters "closed" issues out of milestones.

**Comment by dealingwith** [comment]
@discorick the repo is not public. Here is the meta-data from one of the disappeared tickets in question:

```
<!---
@huboard:{"order":3.87519160848489e-161,"milestone_order":3606,"custom_state":""}
-->
```

**Comment by dealingwith** [comment]
...and yes the ticket still has the column label applied

**Comment by seanlinsley** [comment]
If y'all haven't deployed anything this week, perhaps it's a bug in Github's API.

Yes, this just started happening yesterday.

**Comment by seanlinsley** [comment]
Additional weirdness: my account unassigned people from a couple tickets around the same time yesterday even though I didn't do anything.

<img width="519" alt="screen shot 2016-09-02 at 2 38 44 pm" src="https://cloud.githubusercontent.com/assets/688886/18216204/0306caf8-711b-11e6-920a-ac0afc14c5eb.png">

**Comment by seanlinsley** [comment]
/me resets my password just in case...

**Comment by seanlinsley** [comment]
The disappearance bug actually just happened again, this time for a PR.

I'm not sure whether or not a reload of Huboard is needed for it to disappear.

<img width="585" alt="screen shot 2016-09-02 at 2 42 19 pm" src="https://cloud.githubusercontent.com/assets/688886/18216327/a1594a96-711b-11e6-8e95-c461e604dfa2.png">

**Comment by discorick** [comment]
If you are concerned at all about unauthorized usage of your account I would also revoke all of your OAuth tokens for any GitHub integration you currently use and reauth with each one. 

> The disappearance bug actually just happened again, this time for a PR.

Did the PR disappear on close ?

**Comment by seanlinsley** [comment]
The PR didn't disappear immediately, it was a few minutes later. I'm not sure whether I reloaded Huboard between when it was visible and when it disappeared.

**Comment by seanlinsley** [comment]
I've confirmed that https://huboard.com/api/v2/modernmsg/modernmsg/details is missing the PR in question.

**Comment by seanlinsley** [comment]
It definitely looks like the pubsub is working correctly, but fresh page loads aren't.

**Comment by seanlinsley** [comment]
I just created a test ticket, and opened Huboard in both Chrome and Firefox. For both browsers, the pubsub correctly closes the ticket without archiving it. But after a reload, it's missing from both.

**Comment by discorick** [comment]
Interesting... I am just thinking on this - I did the same thing on a test repo but my issue closed and stayed on the board (even after refresh)

I am not aware of any GitHub API changes - you have more than 5 columns in your board correct ?

**Comment by discorick** [comment]
> I just created a test ticket, and opened Huboard in both Chrome and Firefox. For both browsers, the pubsub correctly closes the ticket without archiving it. But after a reload, it's missing from both.

Is that new issue missing from your details payload ? Could you <s>paste</s> email support at huboard dot com with the meta-data for that one as well ?

**Comment by dealingwith** [comment]
We have 8 columns.

Here is the metadata for that ticket:

```
<!---
@huboard:{"order":3520.352,"milestone_order":3640,"custom_state":""}
-->
```

**Comment by seanlinsley** [comment]
This is quite disruptive to our dev process. Any thoughts on what the issue might be?

Your API server is definitely sending incorrect data (missing these tickets). Whether that's an issue with your server or with Github's API, I don't know.

**Comment by seanlinsley** [comment]
It's a long shot, but I though I'd mention it: around an hour before this bug happened I removed @dealingwith from our Github organization and then added him back.

**Comment by discorick** [comment]
Just exploring this in detail right now @seanlinsley - to make sure I understand correctly
- You started with a Done column with 1 (or more?) issue(s)
- At some point 100 previously archived issues appeared (were they manually archived before or had they simply just ran past the "2 weeks" visible limit?)
- After archiving those issues, only 1 remains in the done column, even though there are issues that have been closed recently and have not been archived

**Comment by discorick** [comment]
Also, do you feel comfortable emailing me the JSON response of:  https://huboard.com/api/v2/modernmsg/modernmsg/details, I have a suspicion of whats going on

**Comment by discorick** [comment]
Also, do you feel comfortable emailing me the JSON response of:  https://huboard.com/api/v2/modernmsg/modernmsg/details, I have a suspicion of whats going on

**Comment by discorick** [comment]
Also, do you feel comfortable emailing me the JSON response of:  https://huboard.com/api/v2/modernmsg/modernmsg/details, I have a suspicion of whats going on

**Comment by seanlinsley** [comment]
> You started with a Done column with 1 (or more?) issue(s)

We started with around 20 closed tickets in Done from our previous sprint.

> At some point 100 previously archived issues appeared (were they manually archived before or had they simply just ran past the "2 weeks" visible limit?)

I'm not sure. They were multiple years old; I don't remember if we were using Huboard back then.

> After archiving those issues, only 1 remains in the done column, even though there are issues that have been closed recently and have not been archived 

Yes, but to clarify that one ticket remaining wasn't closed.

> Also, do you feel comfortable emailing me the JSON response of: https://huboard.com/api/v2/modernmsg/modernmsg/details, I have a suspicion of whats going on

Just sent to support@huboard.com

**Comment by seanlinsley** [comment]
Both the bug I reported, and this ticket seem to be fixed as of the redesign.
